File: haskell/arithmetic.hs
-- | interleave xs ys  
--  return a list of corresponding elements from xs and ys, interleaved. The remaining elements of  
--  the longer list are ignored.  This function should use recursion and pattern matching.  
  
interleave :: [a] -> [a] -> [a]  
interleave (x : xs) (y : ys) = x : y : interleave xs ys  
interleave [] _ = []   
interleave _ [] = []  
  
  
-- | toPairs xs ys  
--  return a list of pairs of corresponding elements from xs and ys, interleaved. The remaining  
--  elements of the longer list are ignored.  This function should use recursion and pattern  
--  matching.  
toPairs :: [a] -> [b] -> [(a, b)]  
toPairs (x : xs) (y : ys) = (x, y) : toPairs xs ys  
toPairs _ [] = []  
toPairs [] _ = []  
  
-- | repeat' f x n  
--  return a list [x, f(x), f(f(x)), ..., f^n(x)]  
--  requires n >= 0  
--  This function should use map, recursion, and pattern matching.  
repeat' :: (a -> a) -> a -> Int -> [a]  
repeat' f x n = x : map (\i -> f (last (repeat' f x (i - 1)))) [1 .. n]  
  
-- i = 0 => x  
-- i = 1 => f last(repeat f x i) => f x  
-- i = 2 => f  
-- repeat f x 0 = [x]  
-- rpeeat f x 1 = [x, f(x)]  
-- repeat f x 2 = [x, f(x), f(fx)]  
  
-- | numNeg xs  
--  return a number of negative elements in xs  
--  No recursion, no higher-order functions. Use list comprehension!  
numNeg :: (Ord a, Num a) => [a] -> Int  
numNeg xs = sum ([1 | a <- xs, a < 0])  
  
-- recursive: if ((head xs) < 0) then 1 + numNeg (tail xs) else numNeg ( tail xs)  
-- numNeg [] = 0  
  
-- | genSquares low high  
--  return a list of squares of even integers between low and high, inclusive.  
--  No recursion, no higher-order functions. Use list comprehension!  
genSquares :: Int -> Int -> [Int]  
genSquares low high = [x * x | x <- [low..high]]  
  
-- | triples n  
--  return a list of triples (x,y,z) of positive numbers all less than  
--  or equal to n, such that x^2 + y^2 == z^2, with no duplicate triples or  
--  premutations of ealier triples.  
--  No recursion, no higher-order functions. Use list comprehension!  
triples :: Int -> [(Int, Int, Int)]  
triples n =  
  [(x, y, z) | x <- [1 .. n],y <- [x .. n], z <- [y + 1 .. n], x^2 + y^2 == z^2] 

File: arithmetic.py
"""Exercise 3."""  
  
  
def interleave(lst1, lst2):  
    """Return a list of corresponding elements from lst1 and lst2, 
    interleaved. The remaining elements of the longer list are ignored. 
 
    This function should use zip and list comprehension, and no recursion. 
 
    >>> interleave([1, 2, 3], ['a', 'b', 'c']) 
    [1, 'a', 2, 'b', 3, 'c'] 
 
    """  
    #zip = [(,), (,)]  
    return [element for pair in zip(lst1, lst2) for element in pair]  
    # [element (For pair in zip (lst1,lst2)), for elemetn in pair]  
    # [pairs from zip, and for every element in pair, want element]  
   
  
def to_pairs(lst1, lst2):  
    """Return a list of pairs of corresponding elements from lst1 and lst2, 
    interleaved. The remaining elements of the longer list are ignored. 
 
    This function should use list comprehension and no recursion. 
 
    >>> to_pairs([1, 2, 3], ['a', 'b', 'c']) 
    [(1, 'a'), (2, 'b'), (3, 'c')] 
    """  
    x = min(len(lst1), len(lst2))  
      
    return [(lst1[i], lst2[i]) for i in range(0,x)]  
   
  #if no zip then...  
  #else zip  
  
def repeat(func, value, num):  
    """Return a list 
    [value, func(value), func(func(value)), ..., func ^ num(value)]. 
    Pre: n >= 0 
 
    This function should use map and recursion. 
 
    >>> repeat(lambda x: 2 * x, 3, 5) 
    [3, 6, 12, 24, 48, 96] 
    """  
  
    if num == 0:  
        return [value]  
    else:  
        return [value] + repeat(func, func(value), num-1)  
  
  
def num_neg(lst1):  
    """Return a number of negative elements in lst1. 
 
    This function should use list comprehension, no recursion, and no 
    map or higher-order functions. 
 
    >>> num_neg([1, -2, 3, -4, -5]) 
    3 
 
    """  
  
    return len([x for x in lst1 if x<0])  
  
  
def gen_squares(low, high):  
    """Return a list of squares of even integers between low and high, inclusive. 
 
    This function should use list comprehension, no recursion, and no 
    map or other higher-order functions. 
    """  
  
    return [(e*e) for e in range(low, high+1) if e%2 == 0]  
  
  
def triples(num):  
    """Return a list of triples (x, y, z) of positive numbers all less 
    than or equal to num, such that x ^ 2 + y ^ 2 == z ^ 2, with no 
    duplicate triples or premutations of ealier triples. 
 
    This function should use list comprehension, no recursion, and no 
    map or other higher-order functions. 
 
    Pre: num > 0. 
 
    >>> triples(5) 
    [(3, 4, 5)] 
 
    """  
  
    return [(x,y,z) for x in range(1,num+1) for y in range(x,num+1) for z in range(y+1, num+1) if x*x + y*y == z*z ]  
  
  
if __name__ == '__main__':  
    import doctest  
    doctest.testmod()  
